Add a time-based STF system test

Mesbah-Alam opened the pull request:
- Adds a time-based STF system test

Related to https://github.com/AdoptOpenJDK/openjdk-tests/issues/2104

FYI @lumpfish 

Signed-off-by: Mesbah_Alam@ca.ibm.com <Mesbah_Alam@ca.ibm.com>

Mesbah-Alam commented:
Test : InternalGrinder/12632

llxia commented on openjdk.test.load/src/test.load/net/adoptopenjdk/stf/TimedLoadTest.java:
```diff
@@ -0,0 +1,92 @@
+	String timelimitInMins;
+	int testCountMultiplier;
+	public void help(HelpTextGenerator help) throws StfException {
+		help.outputSection("MixedLoadTest runs workloads of math,mauve,nio,lang,concurrent unit tests");
```
Is this `TimedLoadTest ` or `MixedLoadTest `?

Mesbah-Alam replied:
Good catch! That was a copy paste error. I've updated the help message.

lumpfish commented:
I do not think 'TimedLoadTest' is what this test should be called - assuming you might want to create other workload tests which are timed based what would you call the next one?

As far as I can see, this is actually just a variant of MixedLoadTest running for a period of time rather rather than a fixed number of tests.  So why not just modify MixedLoadTest to accept the timeLimit argument?  If the argument is not received, then the test would behave as is does now; if the argument is present it would be used as the timeLimit and the number of tests to be run increased so that they do not complete before the timelimit expires.

Also, why only allow the user to specify the timeLimit in minutes?  The STF -timeLimit option actually uses a 5s / 5m / 5h syntax so it makes sense for the -test-args argument to use the same syntax.

If that were done then a playlist or test makefile could have several targets for the same workload such as:
```
MixedLoadTest:
	stf.pl -test=MixedLoadTest
MixedLoadTest_5secs:
	stf.pl -test=MixedLoadTest -test-args="timeLimit=5s"
MixedLoadTest_5mins:
	stf.pl -test=MixedLoadTest -test-args="timeLimit=5m"
MixedLoadTest_5hrs:
	stf.pl -test=MixedLoadTest -test-args="timeLimit=5h"
```

lumpfish reviewed (changes requested):
See comment in PR Conversation regarding requested changes.

Mesbah-Alam commented:
@lumpfish - Thanks for the review comments. The following changes have been made based on them: 
1) Removed TimedLoadTest altogether. 
2) Updated MixedLoadTest to handle the case where the input parameters are passed in - in which case it will set the time cap, otherwise the default test load will be run
3) Updated the [playlsit](https://github.com/AdoptOpenJDK/openjdk-tests/pull/2107/files) to pass in value like `5m` as accepted by STF -timeLimit option. 

Tested time-bsased case here: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5483/ 
Tested default case here: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5486/console

lumpfish commented on openjdk.test.load/src/test.load/net/adoptopenjdk/stf/MixedLoadTest.java:
```diff
@@ -23,14 +23,31 @@
+		//   (1) timelimitInMins     : Time duration (e.g. 5s, 5m, 5h) for which the load should be run. 
+		//   (2) testCountMultiplier : Optional multiplier value to increase the workload externally.
+		StfTestArguments testArgs = stf.env().getTestProperties("timelimitInMins=[0m]","testCountMultiplier=[1]");
+		timelimitInMins = testArgs.get("timelimitInMins");
```
Since the user can now spcify any time units (5ms, 5s, 5m or 5h) the variable should not have the 'InMins' ending.

Mesbah-Alam replied:
Makes sense. Updated the PR.

lumpfish commented on openjdk.test.load/src/test.load/net/adoptopenjdk/stf/MixedLoadTest.java:
```diff
@@ -63,13 +80,25 @@ public void execute(StfCoreExtension test) throws StfException {
+		
+		loadTestInvocation = loadTestInvocation
+				.setSuiteInventory(inventoryFile) 	// Point at the file which lists the tests
+				.setSuiteRandomSelection(); 		// Randomly pick the next test each time
```
The only values different for a time limited test are the time limit and the number of tests.  It would be easier to follow if only `setTimeLimit` and `setSuiteNumTests` were outside the initlal `loadTestInvocation` definition.

Mesbah-Alam replied:
I had tried that, but there was a runtime error from STF:  

```
GEN stderr Exception in thread "main" net.adoptopenjdk.stf.StfException: Java invocation built out of sequence. LOAD_TEST_ARGS cannot be set after SUITE
```

That's why the sequence was kept as it is.

lumpfish commented on openjdk.test.load/src/test.load/net/adoptopenjdk/stf/MixedLoadTest.java:
```diff
@@ -63,13 +80,25 @@ public void execute(StfCoreExtension test) throws StfException {
+				.setSuiteRandomSelection(); 		// Randomly pick the next test each time
 				
 		test.doRunForegroundProcess("Run mixed unit tests", "LT", Echo.ECHO_ON,
 				ExpectedOutcome.cleanRun().within(timeout), 
```
`timeout` is hard coded to `3h`.  It should be adjusted for a time limited test.  The time limit is when stf stops submitting new tests, so they will need time to complete.  To achieve roughly the same as currently behavior `timeout` should be `timeout + timelimit`.

Mesbah-Alam replied:
Is this test ever going to be used to run a load that's 3 hour long (or longer)? The original requirement was to write a STF load test that runs for 5 minutes. That's why the original `3h` timeout was left as is. 

If we plan to run really large loads such as 3h, we need to first make sure that the workload set would last that long.  (the current workload here was only tested to last 20 minutes. It may last much longer-- but I haven't tested it past the 20 minute mark). 

That said, I've updated the test to use a timeout value of `default timeout (3h) + timeLimit`

Mesbah-Alam commented:
Latest changes were tested here: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5489/

lumpfish commented on openjdk.test.load/src/test.load/net/adoptopenjdk/stf/MixedLoadTest.java:
```diff
@@ -22,15 +22,34 @@
+		testCountMultiplier = Integer.parseInt(testArgs.get("testCountMultiplier"));
+		
+		// When We are using MixedLoadTest for default cases;. e.g. not in a time based load test scenario 
+		if ( timeLimit.equals("0m") && testCountMultiplier == 1) {
```
Since you are using the default value to test whether timlimit was specified, the default value could be more descriptive (`unspecified` or `notset` or something?).

Mesbah-Alam replied:
Done. Test: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5490/console

lumpfish commented on openjdk.test.load/src/test.load/net/adoptopenjdk/stf/MixedLoadTest.java:
```diff
@@ -63,16 +80,37 @@ public void execute(StfCoreExtension test) throws StfException {
+				.setSuiteNumTests(totalTests * 80000000 * testCountMultiplier); // If it's a time based test, run a very large load
+		} else {  
+			loadTestInvocation = loadTestInvocation
+				.setSuiteNumTests(totalTests * 200);	// Leave 1 cpu for the JIT. i for GC and set min 20
```
Comment is wrong (looks like it was wrong before this PR - it should be on the `.setSuiteThreadCount(cpuCount - 2, 10)` line, and should say `min 2` not `min 20`.

Mesbah-Alam replied:
Yes, this comment was wrong from before. Thanks for catching! Updated.

lumpfish reviewed (approved):
Looks good.  Please "Squash and Merge" when you're ready.

fjeremic commented on openjdk.test.load/src/test.load/net/adoptopenjdk/stf/MixedLoadTest.java:
```diff
@@ -63,16 +80,37 @@ public void execute(StfCoreExtension test) throws StfException {
+		loadTestInvocation = loadTestInvocation.setAbortIfOutOfMemory(false)
+				.addSuite("mix")
+				.setSuiteThreadCount(cpuCount - 2, 10); // Leave 1 cpu for the JIT, 1 cpu for GC and set min 10
+		if (isTimeBasedLoadTest) { 
+			loadTestInvocation = loadTestInvocation
+				.setSuiteNumTests(totalTests * 80000000 * testCountMultiplier); // If it's a time based test, run a very large load
```
@Mesbah-Alam shouldn't we just set this to `INT_MAX` or `LONG_MAX` or whatever the maximum representable integer is rather than using a "very large load" with a magic 80,000,000 constant? The tests will be terminated after the alloted time anyway.

Mesbah-Alam replied:
Thanks for your suggestion. I will try this out.
